core/templates/dev/head/components/SolutionEditorDirective.js

```diff
@@ -17,63 +17,118 @@
-          $scope.solutionCorrectAnswerEditorHtml = '';
 
           $scope.EXPLANATION_FORM_SCHEMA = {
             type: 'html',
             ui_config: {}
           };
 
           $scope.openSolutionEditor = function() {
-            if ($scope.isEditable) {
-              $scope.solutionEditorIsOpen = true;
-            }
-          };
+            $modal.open({
+              templateUrl: UrlInterpolationService.getDirectiveTemplateUrl(
+                '/pages/exploration_editor/editor_tab/add_solution_modal.html'),
+              backdrop: 'static',
+              controller: [
+                '$scope', '$modalInstance', 'stateInteractionIdService',
+                'stateSolutionService', 'editorContextService',
+                'oppiaExplorationHtmlFormatterService',
+                'explorationStatesService',
+                function($scope, $modalInstance, stateInteractionIdService,
+                         stateSolutionService, editorContextService,
+                         oppiaExplorationHtmlFormatterService,
+                         explorationStatesService) {
+                  $scope.SOLUTION_EDITOR_FOCUS_LABEL = (
+                    'currentCorrectAnswerEditorHtmlForSolutionEditor');
+                  $scope.correctAnswer = null;
+                  $scope.correctAnswerEditorHtml = (
+                    oppiaExplorationHtmlFormatterService.getInteractionHtml(
+                      stateInteractionIdService.savedMemento,
+                      /* eslint-disable max-len */
+                      explorationStatesService.getInteractionCustomizationArgsMemento(
+                        editorContextService.getActiveStateName()),
+                      /* eslint-enable max-len */
+                      $scope.SOLUTION_EDITOR_FOCUS_LABEL));
+                  $scope.EXPLANATION_FORM_SCHEMA = {
+                    type: 'html',
+                    ui_config: {}
+                  };
 
-          $scope.submitAnswer = function(answer) {
-            // This function sets correctAnswer. correctAnswerEditorHtml calls
-            // this function when an answer is input.
-            stateSolutionService.displayed.setCorrectAnswer(answer);
-          };
+                  $scope.data = {
+                    answerIsExclusive: (
+                      stateSolutionService.savedMemento.answerIsExclusive),
+                    correctAnswer: null,
+                    explanation: stateSolutionService.savedMemento.explanation
+                  };
 
-          $scope.saveThisSolution = function() {
-            $scope.solutionEditorIsOpen = false;
-            $scope.getOnSaveFn()();
-          };
+                  $scope.submitAnswer = function(answer) {
+                    $scope.data.correctAnswer = answer;
+                  };
 
-          $scope.cancelThisSolutionEdit = function() {
-            $scope.solutionEditorIsOpen = false;
-          };
+                  $scope.saveSolution = function() {
```

Comment:
Maybe add some light validation here to ensure that correctAnswer is not null and explanation is non-empty and answerIsExclusive is a bool. This helps to provide some defence-in-depth.